### PR TITLE
[refactor] Add Assert, BoolOp, NamedExpr and dict to the new AST builder

### PR DESCRIPTION
**Pull request by lin-hitonami**
Related issue = #3038

### CODE REVIEW

**Review by ailzhang — APPROVED** [review]
LGTM!

File: tests/python38/test_ast_refactor.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from sys import version_info
+
+import pytest
+
+import taichi as ti
+
+
+@ti.test(experimental_ast_refactor=True)
```
Comment (ailzhang) [inline tests/python38/test_ast_refactor.py:8]: can we add a note here to explain why this is a separate file/link to the github issue? :P

**Review by ailzhang — COMMENTED** [review]
inline comment can be a followup PR ;)

**Review by strongoier — APPROVED** [review]
LGTM!

File: tests/python38/test_ast_refactor.py (hunk 1 above)
Comment (lin-hitonami) [reply tests/python38/test_ast_refactor.py:8]: I will add it in the next pr.

**Pull request merged**